ci : Docker 도입 및 새로운 브랜치 전략 제안

ahtop00 opened the pull request:
## 🔎 Description
- 팀 규모(3인)와 배포 안정성을 고려하여 Main-Release 브랜치 전략을 도입하고, 이에 맞춰 CI/CD 파이프라인을 구축했습니다.
- 기존의 단일 브랜치 배포 방식에서 벗어나, main 브랜치는 통합 및 빌드 검증(Staging) 역할을 수행하고, release 브랜치는 실제 서비스 배포(Production) 역할을 담당하도록 분리했습니다. 이를 통해 검증되지 않은 코드가 실서버에 배포되는 사고를 방지합니다.

## 🐳 Docker Build & Deploy Process
> 이번 PR을 통해 Docker 컨테이너 기반의 배포 환경으로 완전히 전환하였습니다. 주요 변경 사항과 동작 방식은 다음과 같습니다.

### 1. 브랜치별 이미지 태그 전략 (Multi-Tagging)
- GitHub Actions에서 현재 브랜치를 감지하여 Docker 이미지 태그를 동적으로 부여합니다.
    - main 브랜치: caring-back:dev 태그로 빌드됩니다. (EC2 배포 X, 빌드 정합성 테스트용)
    - release 브랜치: caring-back:latest 태그로 빌드되며, 실제 EC2 서버에 배포됩니다.

### 2. 배포 파이프라인 (Workflow) 상세
- release 브랜치에 푸시될 경우, 다음 순서로 배포가 진행됩니다.

#### Config Injection (보안)
- 민감한 환경변수와 인증서는 코드 베이스에서 제외하고, safori-team/Caring-Back-Env 리포지토리에서 배포 시점에 안전하게 가져옵니다.
- 가져온 설정 파일들은 압축 및 인코딩되어 AWS SSM을 통해 EC2로 전송됩니다.

#### Docker Build & Push
- Dockerfile을 기반으로 이미지를 빌드하고 Docker Hub에 푸시합니다.

#### AWS SSM Deployment (Zero-Access)
- EC2에 SSH 접속을 하지 않고, **AWS Systems Manager (SSM)**를 통해 명령을 전달합니다.
- 실행 명령:
    - 최신 설정 파일(docker.env, 인증서)을 /home/ubuntu/caring-voice/credentials에 배치
    - 기존 컨테이너 중단 및 docker pull
    - docker run으로 컨테이너 재시작 (Volume Mount로 설정 파일 주입)

### 3. 기대 효과
- 환경 일관성: 로컬, 개발, 운영 환경에서 동일한 Docker 이미지를 사용하여 "내 컴퓨터에선 되는데?" 문제 해결
- 배포 자동화: 수동 SSH 접속 없이 클릭 한 번(Merge)으로 안전하게 배포 가능
- 보안 강화: 환경 변수와 코드를 분리하고, IAM Role 기반으로 권한 제어

## 🔗 Related Issue
- #24 

## 🏷️ What type of PR is this?
- [x] ✨ Feature
- [ ] ♻️ Code Refactor
- [ ] 🐛 Bug Fix
- [ ] 🚑 Hot Fix
- [x] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] ⚡️ Performance Improvements
- [ ] ✅ Test
- [x] 👷 CI
- [ ] 💚 CI Fix

## 📋 Changes Made
- [x] Main-Release 브랜치 전략 적용
    - main: 개발 통합 및 Docker 빌드 테스트 (Tag: :dev)
    - release: 실서버 배포 트리거 (Tag: :latest)

- [x] GitHub Actions Workflow 재구축
    - 브랜치(main/release)에 따라 Docker 이미지 태그를 동적으로 분리
    - release 브랜치 푸시 시에만 AWS SSM을 통해 EC2 배포 스크립트 실행

- [x] 보안 강화 (Config Repo 연동)
    - 민감한 환경변수(.env, 인증서)를 외부 비공개 리포지토리(safori-team/Caring-Back-Env)로 분리
    - 배포 시점에 GH_PAT를 이용해 안전하게 주입받도록 설정 (현재 제 계정의 Token으로 되어있습니다)

## 🧪 Testing
- [x] 테스트용 docker 브랜치 푸시 시 Docker 빌드 정상 동작 확인 및 배포 확인 (이후 마지막 커밋에서 main, release 로 바꾸었습니다)

## 📸 Screenshots (if applicable)
<img width="1540" height="652" alt="image" src="https://github.com/user-attachments/assets/b56469e1-5a54-4011-8ea3-966138edcc39" />  

- Lambda 서버가 있는 AWS 계정에 빈 EC2를 설치하고 내부에서 docker를 설치 후 진행하였습니다.

<img width="1823" height="765" alt="image" src="https://github.com/user-attachments/assets/42e87f76-cf53-43ce-82ad-63c10fd402be" />  

- Health Check를 성공함을 확인하였습니다.
- 현재는 DB 설정이 되어 있지 않기 때문에 준혁님 PR 이후 추가로 병합하여 검증하겠습니다.

<img width="990" height="490" alt="image" src="https://github.com/user-attachments/assets/1545a83f-1b36-4189-916f-b2005d22a30a" />  

- 현재 workflow 제어를 위한 Secret 값입니다. 만약 정한님 계정으로 EC2를 옮긴다면 `AWS_ROLE_ARN` 부분만 변경하면 되어서, 유동성과 확장성 좋습니다! (이후 EC2에 tag를 알맞게 달면 SSM이 자동 인식 후 배포)
- `AWS_ROLE_ARN` 관련은 이후 정한님과 준혁님도 이해하기 쉽게 글로 따로 작성해 알려드리겠습니다! 🙇 

## 📝 Additional Notes
**[브랜치 전략 변경 배경]**
> 기존에는 브랜치 관리가 명확하지 않고, 검증되지 않은 코드가 `main`에 병합되어 배포 안정성이 떨어지는 문제가 있었습니다. 특히 실서버 배포 후 문제가 발생하면 **'수정 → 재배포 → 대기'**의 과정이 반복되어 개발 효율이 저하되었습니다.

- **`main`**: Staging(검증) 역할. 충분한 테스트를 거치는 공간으로 활용합니다.
- **`release`**: Production(실서버) 역할. 검증된 코드만 배포하여 안정성을 확보합니다.

branch 전략을 `feat`(개발) → `main`(검증) → `release`(배포) 3단계로 단순화하여, 관리의 직관성과 배포의 안정성을 모두 확보할 수 있을 것으로 기대됩니다! 검토하시고 괜찮은지 의견 부탁드립니다!

```mermaid
graph LR
    subgraph "1. Local Development (개발)"
        direction TB
        DEV1[🧑‍💻 Dev A] --- FEAT1(feat/login)
        DEV2[🧑‍💻 Dev B] --- FEAT2(feat/api)
        DEV3[🧑‍💻 Dev C] --- FEAT3(feat/fix)
        style FEAT1 stroke-dasharray: 5 5
        style FEAT2 stroke-dasharray: 5 5
        style FEAT3 stroke-dasharray: 5 5
    end

    subgraph "2. Staging / Verification (검증)"
        FEAT1 -->|"PR & Squash Merge"| MAIN{✅ main Branch}
        FEAT2 -->|"PR & Squash Merge"| MAIN
        FEAT3 -->|"PR & Squash Merge"| MAIN
        
        MAIN --"Push Trigger"--> CI_MAIN["🏗️ CI Action (검증용)<br/>Docker Build (Tag: :dev)"]
    end

    subgraph "3. Production Deployment (배포)"
        MAIN --"🚀 배포 시점 PR & Merge"--> RELEASE{🚢 release Branch}
        
        RELEASE --"Push Trigger"--> CD_RELEASE["🏗️ CD Action (배포용)<br/>Docker Build (Tag: :latest)"]
        CD_RELEASE --"🔐 Config 주입"--> SSM["⚡ AWS SSM Deploy Command"]
        SSM -->|"Docker Pull & Restart"| EC2["🖥️ EC2 Live Server<br/>(Production)"]
    end

    %% 관계 설명
    CI_MAIN -.->|"빌드 성공 확인 (배포 X)"| MAIN
```

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * 컨테이너 런타임 설정(포트·환경변수) 및 앱 실행 엔트리 추가
  * 채팅 분석 호출을 위한 서비스 주입 기능 도입

* **Chores**
  * 브랜치 기반 빌드·푸시 및 자동 배포를 포함한 CI/CD 워크플로우 추가(수동 트리거·상태 모니터링)
  * 도커 빌드 컨텍스트에서 일부 구성 파일 제외 해제

* **Dependencies**
  * 멀티파트 처리 패키지 추가 및 대형 모델/오디오용 별도 의존성 묶음 도입

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ahtop00 committed:
ci: add Dockerfile and GitHub Actions workflow for CI/CD deployment

ahtop00 committed:
ci: docker 설정 관련 추가

ahtop00 committed:
ci : 배포 스크립트 로그인 방식 개선 및 로그 추가

- 기존 docker login 방식이 대기 상태로 멈추는 문제 해결
- 보안 강화를 위해 password-stdin 옵션 사용
- 배포 진행 상황 확인을 위한 단계별 로그 추가

ahtop00 committed:
fix: 누락된 라이브러리 추가

ahtop00 committed:
merge origin/main

ahtop00 committed:
Merge branch 'main' into docker

ahtop00 committed:
Merge branch 'main' into docker

ahtop00 committed:
ci : 외부 Config 리포지토리 연동 및 배포 파이프라인 재구축

ahtop00 committed:
ci : 압축 관련 배포 파이프라인 수정

ahtop00 committed:
fix : SSM 명령어 오류 관련 수정

ahtop00 committed:
fix : SSM 명령어 오류 관련 수정 및 태그로 배포

ahtop00 committed:
chore : main, release 브랜치 분기 추가

ahtop00 committed:
chore : 해당 PR 검증을 위한 docker 브랜치 ci/cd 추가

jpark0506 commented:
> @jpark0506 준혁님 제가 어제 잘못 설명해드린 부분이 있어 정정합니다!
> 
> PR을 작성하면서 다시 확인해보니 제가 설명을 잘못 드렸네요ㅠㅡㅠ
> 
> 깃허브 액션에서 이미지는 순수 코드만으로 빌드하고, .env 파일은 배포할 때 EC2 서버에 따로 전송해서 컨테이너 띄울 때 연결(Mount)하는 방식입니다. 즉 빌드시 주입이 아닌 런타임 주입이 맞습니다! 🙇 

아하 그렇군요? 그럼 docker 이미지 생성시에는 필요가 없겠네용 다행입니다. 빌드때 넣는건 까다로워지더라고요

jpark0506 commented:
## 제안해주신 브랜치 전략에 대한 의견

dev 서버가 따로 존재하는게 아니라면, 말씀해주신 것처럼 간단하게 유지해도 무방할 것 같습니다.
다만 여러개를 한번에 머지하게 될 경우, 혹은 hotfix를 해야할 경우 이전 CI / CD를 기다려야 한다는 점이 조금 걱정되긴 합니다.
그래서 dev에서 안정적으로 병합하고, main으로 머지해 빌드 테스트를 해보는 것도 괜찮아 보여요. 각 브랜치의 책임을 나눌 수 있다는 것도 장점이 될 것 같아요. 하지만 말씀드린 건 규모에 비해 살짝(?) 과한 느낌이 있긴 해서 고민이 되네요.

## Docker 빌드 관련

### 빌드 시간 개선

github action 로그를 보았을 때, 빌드 대부분의 시간을 모델 라이브러리를 다운 받는데 사용하는 것 같아요. 해당 시간만 줄이면 더 빠른 빌드가 가능할 것 같습니다. 제가 예전에 해당 기능을 사용해야 했을 때 어려웠던 부분들 공유드립니다.

1. 변경점을 잘못 설정하게 되면, 사소한 변경에도 cache 무효 및 다시 다운로드를 실행합니다. -> 따라서 전용 빌드 레이어를 나누는 것도 추천드립니다.
2. cache를 저장하는데 시간을 꽤 많이 쓰게 됩니다. -> 다운 받는 시간이나, 캐시하는 시간이나 비슷할 수도 있습니다. cache가 되는데 시간이 오래걸린다면, 같은 cache를 또 저장하는지 확인해야 합니다.

지금은 github cache 말고 생각나는 방안은 없긴 한데, 해결하시면 생산성 및 네트워크 비용 감소에 크게 도움이 될 것 같습니다. [관련 레퍼런스](https://tech.kakaoent.com/front-end/2022/220414-docker-cache/)는 FE 밖에 없어서 찾아보고 드리겠습니다.

인스턴스 수준에서는 layer가 재활용 될 것 같아 크게 문제 없을 것 같습니다.

### SSM에 관하여

- 로드밸런싱을 위해서 도입하신 것 같은데, 좋은 것 같습니다. 
- 조금 걱정되는 부분은 다음 옵션들입니다. 1대일때는 상관 없지만, 로드밸런싱 때 다운타임이 크게 발생하지 않을까 걱정되어 조정을 해보면 좋지 않을까 싶습니다.
```
             --max-concurrency "100%" \
            --max-errors "0" \
```

밤에 꽤 고생하신 것 같은데, 훌륭하십니다. 고생하셨습니다!

jpark0506 reviewed (approved):
의견은 코멘트로 남김.

ahtop00 committed:
Merge branch 'main' into docker

ahtop00 committed:
ci : 도커 빌드 최적화 (requirements 분리)

- torch, transformers 등 무거운 라이브러리를 requirements-heavy.txt로 분리
- Dockerfile에서 무거운 패키지를 먼저 설치하여 레이어 캐시 활용
- 빌드 시간 및 재배포 속도 개선

ahtop00 committed:
ci : SSM 오류 수정

ahtop00 committed:
ci : SSM 상태 확인 로직 추가

ahtop00 committed:
ci : OpenSSL 암호화 로직 추가를 통한 보안 개선

ahtop00 committed:
ci : SSM 배포 대기 로직 개선

- SSM 명령 전송 후 성공 여부를 주기적으로 확인하는 로직 추가

ahtop00 committed:
ci : 배포 파이프라인 보안 강화 및 안정성 개선

- CodeRabbit 보안 피드백 반영 (복호화 키 노출 방지 로직 적용)
- SSM 배포 시 무한 대기(Timeout) 현상 해결을 위한 Docker 로그 및 종료 처리 개선

ahtop00 committed:
fix : 누락된 AnalyzeChatService 의존성 import 추가

ahtop00 committed:
fix : API GW 사용시 스테이지 추가로 인해 생기는 문제 수정
